src/array.cr

```diff
@@ -1561,7 +1561,15 @@ class Array(T)
   # b # => [3, 2, 1]
   # a # => [3, 1, 2]
   # ```
-  def sort(&block : T, T -> Int32) : Array(T)
+  def sort(&block : T, T -> U) : Array(T) forall U
+    # TODO: use a better way to check U < Int32?
+    {% begin %}
+      {% block_type = U.union? ? U.union_types.first { |t| t != Nil } : U %}
```

Comment:
Can we add a `TypeNode#contains_type` method? Seems like a simple addition and would help this PR a lot.

Reply:
Isn't `TypeNode#<=` is what you are looking for?

Reply:
Yes, but `Int32?` doesn't work in macros. It's https://github.com/crystal-lang/crystal/issues/6576

I'll try to fix it later. But even if I fix it, I won't be able to use it until the next version (but this PR can wait).

Reply:
I send #6617

With it, we can simply do `{% raise "..." unless U <= Int32? %}`, which is sooo much simpler :-)

Reply:
I didn't know `<=` behaved correctly with union types :)

Reply:
Should we create 0.28 milestone to keep track of this then?

Reply:
I think it's the same code that's used in `is_a?`. There are many related methods that do more or less the same but eventually we should unify them. I now think improving stuff step by step is possible and much better than rewriting many parts.

Reply:
Note: This can be refactored now since 0.27 has the fix

Reply:
@bcardiff You are right! Updated :-)